excel.go

```diff
@@ -0,0 +1,217 @@
+package main
+
+import (
+	"github.com/xuri/excelize/v2"
+)
+
+const (
+	user = iota
+	portal
+	previous
+	timestamp
+	numCols        int    = 4
+	automatedSheet string = "PasswordManager"
+)
+
+func getMCFinUsers(f *excelize.File, config *Portal) (map[string]string, error) {
+	rows, err := f.GetRows(config.SheetName)
+	if err != nil {
+		config.errorLog.Printf("failed getting rows from %s in %s", config.SheetName, config.Filename)
+		return nil, err
+	}
+
+	users := make(map[string]string)
+
+	headerToXCoord := getHeaderToXCoord(rows[0])
+	usernameXCoord := headerToXCoord[config.UsernameHeader]
+	passwordXCoord := headerToXCoord[config.PasswordHeader]
+
+	for _, row := range rows[rowOffset:] {
+		// check if row is empty
+		if row == nil {
+			continue
+		}
+		users[row[usernameXCoord]] = row[passwordXCoord]
+	}
+
+	return users, nil
+}
+
+func getPasswordManagerUsers(f *excelize.File, config *Portal) (map[string]string, error) {
+	rows, err := f.GetRows(automatedSheet)
+	if err != nil {
+		config.errorLog.Printf("failed getting rows from %s in %s", automatedSheet, config.Filename)
+		return nil, err
+	}
+
+	passwordManagerUsersToPassword := make(map[string]string)
+	for _, row := range rows[rowOffset:] {
+		passwordManagerUsersToPassword[row[user]] = row[portal]
+	}
+
+	return passwordManagerUsersToPassword, nil
+}
+
+// Sync PasswordManager usernames with MACFIN users
+func syncPasswordManagerUsersToMACFINUsers(f *excelize.File, config *Portal) error {
+	macFinUsersToPasswords, err := getMCFinUsers(f, config)
+	if err != nil {
+		return err
+	}
+
+	passwordManagerUsersToPassword, err := getPasswordManagerUsers(f, config)
+	if err != nil {
+		config.errorLog.Printf("failed reading users from %s", automatedSheet)
+		return err
+	}
+
+	numRows := len(passwordManagerUsersToPassword)
+
+	// add new MACFIN users to automatedSheet
+	for user, password := range macFinUsersToPasswords {
+		if _, ok := passwordManagerUsersToPassword[user]; !ok {
+			values := [numCols]string{user, password, password, "Rotate Now"}
+			f.InsertRow(automatedSheet, numRows+1) // insert before numRows+1
+			numRows++
+			// write row
+			for j := 0; j < numCols; j++ {
+				err := writeCell(f, config.Filename, automatedSheet, j+sheetOffset, numRows, values[j])
+				if err != nil {
+					config.errorLog.Printf("failed adding new macFin user %s to %s sheet in file %s",
+						user, automatedSheet, config.Filename)
+					return err
+				}
+			}
+			continue
+		}
+	}
+
+	// remove deleted MACFIN users from PasswordManager sheet
+	rowIndx := 1
+	for pwUser := range passwordManagerUsersToPassword {
+		if _, ok := macFinUsersToPasswords[pwUser]; !ok {
+			// pwUser is not in MCFIN users; remove pwUser from PasswordManager
+			err := f.RemoveRow(automatedSheet, rowIndx+rowOffset)
```

Comment:
this is a map iteration right? so the users will be in random order, which I think means you will be removing arbitrary rows

Reply:
Yes. This is a bug. Thank you.
In order to delete from the automatedSheet, I do 2 passes. First pass: mark rows for deletion. Second pass: remove marked rows; iterate from bottom row to top.
In addition to mapping usersToPassword, I also map usersToRow. (see last commit)

Reply:
To delete rows correctly, I've changed `getPasswordManagerUser` in commit 7d0b895d81. It now returns 3 values (`usersToRow`, `usersToPassword` and error). Should I make this 2 functions, with each returning 1 map?

`syncPasswordManagerUsersToMACFinUsers` needs `usersToRow`. `updateMACFinUsers` needs `usersToPassword`.